### ISSUE TITLE
VertexOnlyMesh Point Location Tolerance and Supported Mesh Improvements

### ISSUE THREAD

**Issue by ReubenHill** [firedrakeproject/firedrake#2224]
As identified in #2178 , VertexOnlyMesh's point location and rank distribution system has no way of specifying a tolerance to avoid the loss of points on parent mesh cell or mesh boundaries. This is because it relies on [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) which doesn't allow such a tolerance to be set.

To get around this I should use `mesh.locate_cell` as the initial cell-location method rather than relying on `DMSwarmSetPointCoordinates`. This may also have the advantage of allowing high-order (bendy) meshes and meshes made from cell coordinates to be supported.

Note that firedrake cell numbering will still need to be translated to DMPlex cell numbering if I want to keep the underlying DMSWARM updated in PIC mode (which would keep features like DMSwarmMigrate working).

**Comment by ReubenHill** [comment]
Note that different ranks might make different decisions about which cell a point lives in so care will need to be taken that points on cell boundaries don't get duplicated by accident when running in parallel. 

Take the following `test_parallel_point_locate.py` for example
```python
from firedrake import *
m = UnitSquareMesh(1,1)
coords = [[0.5, 0.5], [0.9, 0.9]]
print(f'Rank {COMM_WORLD.rank}, coord {coords[0]} in cell {m.locate_cell(coords[0])}, coord {coords[1]} in cell {m.locate_cell(coords[1])}')
```
when run on 3 ranks we get
```
$ mpiexec -n 3 python test_parallel_point_locate.py
Rank 0, coord [0.5, 0.5] in cell 0, coord [0.9, 0.9] in cell 0
Rank 1, coord [0.5, 0.5] in cell None, coord [0.9, 0.9] in cell None
Rank 2, coord [0.5, 0.5] in cell 0, coord [0.9, 0.9] in cell 1
```
The cell boundary point `[0.5, 0.5]` has been located in the same cell as that containing the point `[0.9, 0.9]` on rank 0 and in a different cell on rank 2.

This will only be an issue when broadcasting the point locations from rank 0 (i.e. taking it that all ranks have been given the same coordinates list as here). Notably it seems that [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) doesn't suffer from this problem so it would be worth checking how the implementation there works.

One way to deal with this would be to give points unique IDs at the point of creation then check if all ranks agree about their cell ID after calling `locate_cell` but it's a lot of extra communication.

**Comment by ReubenHill** [comment]
Note that the fact we, at present, remove points from halo regions does not help with the above problem. If two ranks were to _both_ decide that a point was in a halo region then it would be expunged forever (incorrectly!)

**Comment by ReubenHill** [comment]
> Note that the fact we, at present, remove points from halo regions does not help with the above problem. If two ranks were to _both_ decide that a point was in a halo region then it would be expunged forever (incorrectly!)

This can be resolved with a voting algorithm which makes sure only one process owns each uniquely defined point.

**Comment by ReubenHill** [comment]
> This will only be an issue when broadcasting the point locations from rank 0 (i.e. taking it that all ranks have been given the same coordinates list as here). Notably it seems that [DMSwarmSetPointCoordinates](https://petsc.org/release/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html) doesn't suffer from this problem so it would be worth checking how the implementation there works.

This appears to be incorrect - see discussion #2450

**Comment by ReubenHill** [comment]
At present, only high order meshes and extruded meshes with variable layers are unsupported.

**Comment by ReubenHill** [comment]
Also the point location tolerance issue has been solved in #2790
